src/url.ts

```diff
@@ -96,14 +96,61 @@ export class $URL implements URL {
     Object.assign(this.query, url.query)
 
     if (url.pathname) {
-      this.pathname = withTrailingSlash(this.pathname) + withoutLeadingSlash(url.pathname)
+      this.pathname = joinURL(this.pathname, url.pathname)
     }
 
     if (url.hash) {
       this.hash = url.hash
     }
   }
 
+  resolve (url: $URL) {
+    this.hash = url.hash
+
+    if (Object.keys(url.query).length || url.pathname) {
+      this.query = url.query
+    }
+
+    if (!url.pathname) {
+      return
+    }
+
+    if (url.hasProtocol) {
+      this.protocol = url.protocol
+      this.host = url.host
+      this.auth = url.auth
+      this.pathname = url.pathname
+      this.query = url.query
+      this.hash = url.hash
+      return
+    }
+
+    if (isAbsolutePath(url.pathname)) {
+      this.pathname = url.pathname
+      return
+    }
+
+    const wasAbsolute = isAbsolutePath(this.pathname)
+
+    const segments = [
```

Comment:
I would suggest moving resolve/segmentation algorithm to a util so that we can also support it in `joinURL`. 

`$URL`/`resolveURL` use full-parsing (encode/decode which includes dependency on punycode) and reason we have `joinURL` is that we can leverage tree-shaking to do lighter join

Reply:
Agree - sounds good. I will separate out the resolve/segmentation algorithm.